zebra-network/src/peer/handshake.rs

```diff
@@ -34,45 +34,120 @@ use super::{Client, Connection, ErrorSlot, HandshakeError};
+    pub fn with_inbound_service(mut self, inbound_service: S) -> Self {
+        self.inbound_service = Some(inbound_service);
+        self
+    }
+
+    /// Provide a hook for timestamp collection. Optional.
+    ///
+    /// If this is unset, timestamps will not be collected.
+    pub fn with_timestamp_collector(mut self, timestamp_collector: mpsc::Sender<MetaAddr>) -> Self {
+        self.timestamp_collector = Some(timestamp_collector);
+        self
+    }
+
+    /// Provide the services this node advertises to other peers.  Optional.
+    ///
+    /// If this is unset, the node will advertise itself as a client.
+    pub fn with_advertised_services(mut self, services: PeerServices) -> Self {
+        self.our_services = Some(services);
+        self
+    }
+
+    /// Provide this node's user agent.  Optional.
+    ///
+    /// This must be a valid BIP14 string.  If it is unset, the user-agent will be empty.
+    pub fn with_user_agent(mut self, user_agent: String) -> Self {
+        self.user_agent = Some(user_agent);
+        self
+    }
+
+    /// Whether to request that peers relay transactions to our node.  Optional.
+    ///
+    /// If this is unset, the node will not request transactions.
+    pub fn want_transactions(mut self, relay: bool) -> Self {
+        self.relay = Some(relay);
+        self
+    }
+
+    /// Consume this builder and produce a [`Handshake`].
+    ///
+    /// Returns an error only if any mandatory field was unset.
+    pub fn finish(self) -> Result<Handshake<S>, &'static str> {
+        let config = self.config.ok_or("did not specify config")?;
+        let inbound_service = self
+            .inbound_service
+            .ok_or("did not specify inbound service")?;
+        let timestamp_collector = self.timestamp_collector.unwrap_or_else(|| {
+            // No timestamp collector was passed, so create a stub channel.
+            // Dropping the receiver means sends will fail, but we don't care.
+            let (tx, _rx) = mpsc::channel(1);
+            tx
+        });
+        let nonces = Arc::new(Mutex::new(HashSet::new()));
+        let user_agent = self.user_agent.unwrap_or_else(|| "".to_string());
+        let our_services = self.our_services.unwrap_or_else(PeerServices::empty);
+        let relay = self.relay.unwrap_or(false);
+        Ok(Handshake {
             config,
-            internal_service,
+            inbound_service,
             timestamp_collector,
-            nonces: Arc::new(Mutex::new(HashSet::new())),
+            nonces,
+            user_agent,
+            our_services,
+            relay,
+        })
+    }
+}
+
+impl<S> Handshake<S>
+where
+    S: Service<Request, Response = Response, Error = BoxedStdError> + Clone + Send + 'static,
+    S::Future: Send,
+{
+    /// Create a builder that configures a [`Handshake`] service.
+    pub fn builder() -> Builder<S> {
+        // can't derive Builder::default without a bound on S :(
```

Comment:
```suggestion
        // We don't derive `Default` because the derive inserts a `where S: Default` bound even tho `Option<S>` implements `Default` even if `S` does not.
```